File: src/connectors/market_data_client.py
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Dict


@dataclass
class Quote:
    symbol: str
    last: float
    bid: float
    ask: float
    volume: int
    timestamp_utc: str


class MarketDataClient:
    def __init__(self, api_key: str, base_url: str) -> None:
        self.api_key = api_key
        self.base_url = base_url

    def get_quote(self, symbol: str) -> Quote:
        """Placeholder adapter. Replace with real market data API integration."""
        ts = datetime.now(timezone.utc).isoformat()
        return Quote(symbol=symbol, last=0.0, bid=0.0, ask=0.0, volume=0, timestamp_utc=ts)

    def get_quotes(self, symbols: list[str]) -> Dict[str, Quote]:
        return {symbol: self.get_quote(symbol) for symbol in symbols}

File: src_cpp/connectors/market_data_client.cpp
#include "connectors/market_data_client.hpp"

#include <chrono>
#include <ctime>
#include <iomanip>
#include <sstream>

namespace {
std::string now_utc() {
    const auto now = std::chrono::system_clock::now();
    const std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm_utc {};
#ifdef _WIN32
    gmtime_s(&tm_utc, &t);
#else
    gmtime_r(&t, &tm_utc);
#endif
    std::ostringstream oss;
    oss << std::put_time(&tm_utc, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}
}

MarketDataClient::MarketDataClient(std::string api_key, std::string base_url, std::string quote_path)
    : api_key_(std::move(api_key)), base_url_(std::move(base_url)), quote_path_(std::move(quote_path)) {}

Quote MarketDataClient::get_quote(const std::string& symbol) const {
    Quote q;
    q.symbol = symbol;
    q.timestamp_utc = now_utc();

    if (base_url_.empty()) {
        return q;
    }

    // Placeholder for real HTTP integration. Replace with parsed fields from market API.
    if (symbol == "NVDA") {
        q.last = 895.0;
    } else if (symbol == "MSFT") {
        q.last = 415.0;
    } else if (symbol == "AAPL") {
        q.last = 195.0;
    } else if (symbol == "TSLA") {
        q.last = 185.0;
    } else {
        q.last = 100.0;
    }

    q.bid = q.last - 0.05;
    q.ask = q.last + 0.05;
    q.volume = 1000000;
    return q;
}
